kep/205-preemption-toleration/README.md

```diff
@@ -0,0 +1,126 @@
+```
+
+### Guarantee to running at least N minutes even in lower priority
+
+This is a typical use-case in machine learning job. As described above, job would need check-pointing. However, most machine learning jobs are iteration-based process, called _epochs_. So, if the cluster is high load and preemption always happens before the first epoch is finished,  the machine learning jobs can never make any progress and computational resources consumed by the job came to nothing in this case.
+
+So, cluster administrators would like to exempt pods that do not finish the first epoch from preemption even if the job has lower priority. But, it is hard to monitor when the first epoch is finished outside of the job. Thus, a minimum running time is useful practically.
+
+For example, cluster administrator would introduce several priority classes with the same priority value for introducing several minimum running time guarantee categories like below:
+
+```yaml
+    system-critical: 10000
+                low:  8000  # no minimum running time guarantee
+          low-10min:  8000  # 10 min minimum running time guarantee against 8000 < p < 10000
+          low-30min:  8000  # 30 min minimum running time guarantee against 8000 < p < 10000
+```
+
+## Design Details
+
+### Preemption Toleration API
+
+Preemption toleration policy is defined in this struct type.
+
+```go
+// Please note that this does NOT have ObjectMeta 
+// because this will be embedded in annotation in PriorityClass as unnamed object.
+type PreemptionToleration struct {
+	metav1.TypeMeta `json:",inline"`
+
+  // MinimumPreemptablePriority specifies the minimum priority value that can preempt this priority class.
+  MinimumPreemptablePriority *int32 `json:"minimumPreemptablePriority,omitempty"`
+
+  // TolerationSeconds specified how many seconds this priority class can tolerate preemption by priorities lower than MinimumPreemptablePriority.  Null value specifies forever (default).  Duration means the duration from the pod being scheduled to some node.  This value affects only to scheduled pods (no effect on nominated nodes).
+  TolerationSeconds *int64 `json:"tolerationSeconds,omitempty"`
+
+  // more policy will arise in the future.
+}
+```
+
+Users specifies the preemption toleration policy as an annotation with the key `scheduling.sigs.k8s.io/preemption-toleration` in `PriorityClass` as below:
+
+```yaml
+kind: PriorityClass
+metadata:
+  name: toleration-policy-sample
+  annotation:
+    scheduling.sigs.k8s.io/preemption-toleration: |
+      # the api is versioned for future enhancement
+      apiVersion: scheduling.sigs.k8s.io/v1beta1
+      kind: PreemptionToleration
+
+      # This priority class can tolerate preemption by priority with 8000 <= p < 10000.
+      minimumPreemptablePriority: 10000
+
+      # And it can tolerate preemption in 1 hour by the pod with priority (8000 <= p < 10000).
+      tolerationSeconds: 3600
```

Comment:
I'd prefer to exact all the settings to a CustomResource object; otherwise, I don't think it's practical to versioning/defaulting/converting on the annotated string.

Or, you give up support multi-version toleration policies - i.e., define the settings as scheduling.sigs.k8s.io/preemption-toleration-XYZ, individually.

Reply:
I see.  If we choose CRD case, `PriorityClass` would need to refer `PreemptionToleration` CR by annotation or a similar way.  Then, I would like to choose the annotation case.

Reply:
done in https://github.com/kubernetes-sigs/scheduler-plugins/pull/205/commits/335c4118c5e2f2010e4d910eb2524ca4abb0e0fb